src/Microsoft.FeatureManagement/FeatureManager.cs

```diff
@@ -96,10 +99,11 @@ private async Task<bool> IsEnabledWithVariantsAsync<TContext>(string feature, TC
             {
                 isFeatureEnabled = false;
             }
-            else if ((featureDefinition.Variants?.Any() ?? false) && featureDefinition.Allocation != null)
-            {
-                VariantDefinition variantDefinition;
 
+            VariantDefinition variantDefinition = null;
+
+            if (featureDefinition != null && (featureDefinition.Variants?.Any() ?? false) && featureDefinition.Allocation != null)
```

Comment:
why'd the else go away? We shouldn't hit both blocks right?

Reply:
If the FeatureStatus is Disabled, we still want to hit the variants block to allocate the disabled variant, if there is any Allocation to use.

Reply:
This is only relevant for telemetry. The result of this call "IsEnabledAsync" can't be changed by variants when the status is disabled (even if there was a StatusOverride) so that's why it doesn't matter to the Variants PR, but does matter to this PR.

Reply:
It makes sense, I think the issue is actually with the presence of the block above. I left a comment on it.

Reply:
My comment on the block above keeps getting swallowed, so I'll leave it here.

The issue is with this piece.

```
            if (featureDefinition == null || featureDefinition.Status == FeatureStatus.Disabled)
            {
                isFeatureEnabled = false;
            }
```

It's basically a duplicate of things we're already figuring out in the IsEnabledAsync call. It looks like we should be doing to avoid it is getting the feature definition before calling IsEnabledAsync, and passing the definition to it.

Reply:
Sorry, I don't follow, which comment?

Reply:
see the previous reply

Reply:
Or, to avoid that big refactor I am fine if we go with this approach

```

        private async Task<bool> IsEnabledWithVariantsAsync<TContext>(string feature, TContext appContext, bool useAppContext, CancellationToken cancellationToken)
        {
            FeatureDefinition featureDefinition = await GetFeatureDefinition(feature).ConfigureAwait(false);

            VariantDefinition variantDefinition = null;

            if (featureDefinition != null)
            {
                bool isFeatureEnabled = await IsEnabledAsync(feature, appContext, useAppContext, cancellationToken).ConfigureAwait(false);
                
                //
                // Get variant definition
            }

            //
            // cache session and publish
        }
```

Reply:
Went with that approach with slight variations for the publishing event at the end.

Reply:
Well, actually I did include the `IsEnabledAsync(..., featureDefinition)` suggestion. Since it's a private method we only call in 2 places, it was easy to include.